src/dht.cpp

```diff
@@ -1341,41 +1398,11 @@ Dht::searchStep(std::shared_ptr<Search> sr)
         while (sent and sr->currentlySolicitedNodeCount() < MAX_REQUESTED_SEARCH_NODES);
         /*DHT_LOG.DEBUG("[search %s IPv%c] step: sent %u requests (total %u).",
             sr->id.toString().c_str(), sr->af == AF_INET ? '4' : '6', i, sr->currentlySolicitedNodeCount());*/
+    }
 
-        auto expiredn = (size_t)std::count_if(sr->nodes.begin(), sr->nodes.end(), [&](const SearchNode& sn) {
-                    return sn.candidate or sn.node->isExpired();
-                });
-        if (i == 0 && expiredn == sr->nodes.size())
-        {
-            DHT_LOG.WARN("[search %s IPv%c] expired", sr->id.toString().c_str(), sr->af == AF_INET ? '4' : '6');
-            // no nodes or all expired nodes
-            sr->expired = true;
-            if (sr->announce.empty() && sr->listeners.empty()) {
-                // Listening or announcing requires keeping the cluster up to date.
-                sr->setDone();
-            }
-            {
-                auto get_cbs = std::move(sr->callbacks);
-                for (const auto& g : get_cbs) {
-                    if (g.second.done_cb)
-                        g.second.done_cb(false, {});
-                }
-            }
-            {
-                std::vector<DoneCallback> a_cbs;
-                a_cbs.reserve(sr->announce.size());
-                for (auto ait = sr->announce.begin() ; ait != sr->announce.end(); ) {
-                    if (ait->callback)
-                        a_cbs.emplace_back(std::move(ait->callback));
-                    if (not ait->permanent)
-                        ait = sr->announce.erase(ait);
-                    else
-                        ait++;
-                }
-                for (const auto& a : a_cbs)
-                    a(false, {});
-            }
-        }
+    if (sr->getNumberOfBadNodes() >= std::min(sr->nodes.size(), static_cast<size_t>(SEARCH_MAX_BAD_NODES))) {
```

Comment:
> The number 25 is based on table 3 from the annex 1 of #137 (k=3/4). Going from 3/4 to 7/8, the number is at least 50. Which seems odd to me.

It makes sense, since if 7/8 nodes are expired, it's then normal to get see many consecutive unreachable nodes, so it takes more successive expired nodes to make sure we are actually disconnected and not just seeing random unreachable nodes.

> If I undestand correctly, you suggest to exclusively consider expired nodes in the node count, right? I also don't count consecutive expired nodes. May be that would be more appropriate.

Current implementation would consider a search with 8 good nodes, then many expired nodes as expired (but it's actually a fully working synced search).
We should count the number of consecutive expired nodes starting from the top of the search.

> I'm not sure I get what you're saying.

If I try to sum it up:
- we chose to consider search expiration as a connectivity loss
- When we detect network connectivity change, all searches should be affected by the expired nodes cleanup, (plus calling connectivityChanged). After this, the state of the DHT must be at best _CONNECTING_ to be consistent, and so that the connectivity change information bubbles up. 
- currently a search expires when all of its nodes are expired
- currently a search has no node count limit, (the good nodes could be anywhere in the node list), so the search expiration time is not bounded
- we can use our stat work results to put an upper limit on the search node count, and the current search expiration criteria can stay the same (it expires when all search nodes are expired). That would put an upper bound to the search expiration time.
- another possibility of implementation is to accept more nodes in the search, and count successive expired nodes, up to Max (our limit for search expiration). But then, what to do with the excessive nodes ? We won't be using them anyway, those nodes are far away, and just before expiring the search we started to send them requests (since they are next in the list after the Max expired nodes).

Yeah it sums it up, but I'm still not sure about how to handle some corner cases (listen recovery...).